classes/auto_login.php

```diff
@@ -0,0 +1,212 @@
+ */
+class auto_login {
+    /** @var bool True if already considered auto-login on this page. */
+    protected static $considered;
+
+    /**
+     * Called on every page to give an opportunity to auto-login.
+     *
+     * This is called on most pages via the after_require_login callback, and also by the
+     * before_http_headers callback, so it might be called twice. We only want to run it once.
+     */
+    public static function consider() {
+        global $SESSION, $SCRIPT;
+
+        if (self::$considered) {
+            return;
+        }
+        self::$considered = true;
+
+        // Don't do this on any auth or login scripts!
+        if (preg_match('~^(/auth/|/login/)~', $SCRIPT)) {
+            return;
+        }
+
+        // If this is not a GET request, don't try autologin.
+        if ($_SERVER['REQUEST_METHOD'] !== 'GET') {
+            return;
+        }
+
+        // If they are already logged in, we don't need autologin.
+        if (isloggedin() && !isguestuser()) {
+            return;
+        }
+
+        // Check if this plugin is enabled.
+        $enabled = \core_plugin_manager::instance()->get_enabled_plugins('auth');
+        if (!array_key_exists('saml2', $enabled)) {
+            return;
+        }
+
+        // Get config.
+        $auth = get_auth_plugin('saml2');
+        switch ($auth->config->autologin) {
+            // Turned off.
+            case saml2_settings::OPTION_AUTO_LOGIN_NO:
+                return;
+
+            // Login once per session. Don't try login if we already tried this session.
+            case saml2_settings::OPTION_AUTO_LOGIN_SESSION:
+                if (!empty($SESSION->auth_saml2_triedautologin)) {
+                    return;
+                }
+                break;
+
+            // Login when a cookie exists or changes. Don't try login if we already tried for this
+            // cookie value, or the cookie isn't set.
+            case saml2_settings::OPTION_AUTO_LOGIN_COOKIE:
+                // If the cookie setting isn't filled in, do nothing.
+                if (empty($auth->config->autologincookie)) {
+                    return;
+                }
+                // If the cookie isn't set, do nothing.
+                if (empty($_COOKIE[$auth->config->autologincookie])) {
+                    return;
+                }
+                // If the cookie is set to the same value it was last time we looked, do nothing.
+                $currentcookie = $_COOKIE[$auth->config->autologincookie];
+                if (!empty($SESSION->auth_saml2_lastautologincookie) &&
+                        $SESSION->auth_saml2_lastautologincookie === $currentcookie) {
+                    return;
+                }
+                break;
+        }
+
+        // Just as a final check (now we're definitely going to try login), is the plugin actually
+        // configured? It probably is since they set up the autologin option, but...
+        if (!$auth->is_configured()) {
```

Comment:
Not a big deal really, but for me it looks like this check should go first before we even go to switch options,.

Reply:
It was after the switch for performance reasons - I was trying to be very careful about performance because this code runs on every page. is_configured only really does a few file_exists checks which is not exactly a big deal, but even so, I think it might as well stay later - I'll add a comment to explain why it is there.